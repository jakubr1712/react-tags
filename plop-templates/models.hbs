export interface I{{properCase name}}ComponentProps {}