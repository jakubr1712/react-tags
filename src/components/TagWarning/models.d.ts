export interface ITagWarningProps {
  currentTags: number;
}