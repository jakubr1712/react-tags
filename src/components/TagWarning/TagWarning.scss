.tag-warning {
  &__container {
    padding: 16px;
    border-radius: 8px;
    margin-top: 16px;
  }

  &__header {
    display: flex;
    align-items: center;
    font-size: 16px;
    font-weight: bold;
    margin-bottom: 8px;
  }
}
