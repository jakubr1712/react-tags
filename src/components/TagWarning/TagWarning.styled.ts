import styled from 'styled-components';

export const WarningContainer = styled.div`
  padding: 16px;
  border-radius: 8px;
  margin-top: 16px;
`;

export const WarningHeader = styled.div`
  display: flex;
  align-items: center;
  font-size: 16px;
  font-weight: bold;
  margin-bottom: 8px;
`;

export const ProgressBar = styled.div`
  flex: 1;
  height: 4px;
  background-color: #ccc;
  border-radius: 2px;
  margin-left: 8px;
  position: relative;
`;

export const ProgressBarFill = styled.div`
  height: 100%;
  border-radius: 2px;
  width: 0;
`;

export const WarningText = styled.div`
  display: flex;
  align-items: center;
  font-size: 14px;
`;

export const WarningIcon = styled.span`
  margin-right: 8px;
  font-size: 18px;
`;
