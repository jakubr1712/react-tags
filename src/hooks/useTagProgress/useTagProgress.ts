import { useState, useEffect } from "react";
import { getProgressColor } from "utils/getProgressColor";
import { getProgressText } from "utils/getProgressText";
import { getWarningMessage } from "utils/getWarningMessage";

export const useTagProgress = (currentTags: number) => {
  const requiredTags = 5;
  currentTags = currentTags >= requiredTags ? requiredTags : currentTags;

  const [progress, setProgress] = useState(0);
  const [color, setColor] = useState("#f44336");
  const [text, setText] = useState("SÅ‚abo");
  const [message, setMessage] = useState("");

  useEffect(() => {
    const progressValue = (currentTags / requiredTags) * 100;
    setProgress(progressValue);
    setColor(getProgressColor(currentTags));
    setText(getProgressText(currentTags));
    setMessage(getWarningMessage(currentTags, requiredTags));
  }, [currentTags]);

  return { progress, color, text, message };
};
