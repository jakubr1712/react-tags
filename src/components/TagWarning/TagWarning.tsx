import React from "react";
import ProgressBar from "../ProgressBar";
import WarningText from "../WarningText";
import { useTagProgress } from "hooks/useTagProgress";
import {ITagWarningProps} from './models.d';
import './TagWarning.scss';

const TagWarning: React.FC<ITagWarningProps> = ({ currentTags }) => {
  const { progress, color, text, message } = useTagProgress(currentTags);

  return (
    <div className="tag-warning__container">
      <div className="tag-warning__header" style={{ color }}>
        {text}
        <ProgressBar progress={progress} color={color} />
      </div>
      <WarningText message={message} color={color} />
    </div>
  );
};

export default TagWarning;
