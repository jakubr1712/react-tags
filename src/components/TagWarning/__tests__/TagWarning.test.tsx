import React from "react";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import TagWarning from "../TagWarning";
import { useTagProgress } from "hooks/useTagProgress";
import { ITagWarningProps } from "../models.d";

jest.mock("hooks/useTagProgress");

describe("TagWarning component", () => {
  const mockUseTagProgress = useTagProgress as jest.MockedFunction<
    typeof useTagProgress
  >;

  const defaultProps: ITagWarningProps = {
    currentTags: 3,
  };

  beforeEach(() => {
    mockUseTagProgress.mockReturnValue({
      progress: 60,
      color: "#cddc39",
      text: "Słabo",
      message:
        "Zbyt mało tagów. Dodaj jeszcze 2 aby poprawić widoczność artykułu",
    });
  });

  it("renders properly and matches snapshot", () => {
    const { container } = render(<TagWarning {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  it("displays the correct progress text and color", () => {
    render(<TagWarning {...defaultProps} />);
    expect(screen.getByText("Słabo")).toBeInTheDocument();
    expect(screen.getByText("Słabo")).toHaveStyle("color: #cddc39");
  });

  it("displays the correct warning message", () => {
    render(<TagWarning {...defaultProps} />);
    expect(
      screen.getByText(
        "Zbyt mało tagów. Dodaj jeszcze 2 aby poprawić widoczność artykułu"
      )
    ).toBeInTheDocument();
  });
});
