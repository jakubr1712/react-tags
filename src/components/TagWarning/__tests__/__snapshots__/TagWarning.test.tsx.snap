// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagWarning component renders properly and matches snapshot 1`] = `
<div>
  <div
    class="tag-warning__container"
  >
    <div
      class="tag-warning__header"
    >
      <span
        style="color: rgb(205, 220, 57);"
      >
        Słabo
      </span>
      <div
        class="progress-bar__container"
      >
        <div
          class="progress-bar__item progress-bar__item--active"
          role="progressbar-item--active"
          style="background-color: rgb(205, 220, 57);"
        />
        <div
          class="progress-bar__item progress-bar__item--active"
          role="progressbar-item--active"
          style="background-color: rgb(205, 220, 57);"
        />
        <div
          class="progress-bar__item progress-bar__item--active"
          role="progressbar-item--active"
          style="background-color: rgb(205, 220, 57);"
        />
        <div
          class="progress-bar__item "
          role="progressbar-item"
        />
        <div
          class="progress-bar__item "
          role="progressbar-item"
        />
      </div>
    </div>
    <div
      class="warning-text"
    >
      <span
        class="warning-text__icon"
      >
        ℹ️
      </span>
      Zbyt mało tagów. Dodaj jeszcze 2 aby poprawić widoczność artykułu
    </div>
  </div>
</div>
`;
